### PR TITLE
Correctly apply quota

### PR DESCRIPTION
**Pull request by jvillafanez**
Expected behaviour with these changes:
* If both LDAP quota field and default LDAP quota are set:
    * if the user has a value for the LDAP quota attribute, try to use that value as quota
    * if the user doesn't have a value for the LDAP quota attribute or the value can't be parsed correctly, use the default LDAP quota.
    * if the user doesn't have a value for the LDAP quota attribute or the value can't be parsed correctly, and the default LDAP quota can't be parsed correctly, use the ownCloud's default quota.
* If the LDAP quota field is empty and the default LDAP quota is set:
    * go through the above case and consider the user doesn't have the LDAP quota attribute set.
* If the LDAP quota field isn't empty and the LDAP default quota isn't set:
    * if the user has a value for the attribute, try to use that value as quota.
    * if the user doesn't have a value for the attribute **do not change anything**. This usually means that the ownCloud's default quota will be used. If the user's quota has a previous value, that value will be kept
* If neither the LDAP quota field nor the default LDAP quota is set, keep the current quota.

Additional changes:
* Explictly support "default" and "none" as valid values for the LDAP quota, both as default LDAP quota and as value for the attribute (not sure if it was supported before).
* Added new logs to troubleshoot weird quota strings.

Notes:
* Even if the user "uses" the default LDAP quota, that value **will be effectively set** as quota. If the admin remove the default LDAP quota at a later time, the previous default LDAP quota might still be used by other users.
* Changes in the quota aren't applied uniformly. They depend on when the user data is fetched from the LDAP. If you set a default LDAP quota value "A", then you change it to "B" and finally change it again to "C", quota values for different users might be "A", "B" or "C" at some point. It's expected that when you access to the user data (or the user logs in) the quota is applied correctly. However, keep in mind that there are some cases (check above) where the quota isn't changed, which will result in unexpected values for the quota.

@PVince81 @DeepDiver1975 to review
@owncloud/qa to verify and check the behaviour
@settermjd in case there is docs to be updated / improved due to this change

### CODE REVIEW

**Comment by jvillafanez** [comment]
This should be enough for now. I plan to add some additional tests to cover as much cases as possible, going along with the current tests.

**Comment by davitol** [comment]
:+1: WFM

> Changes in the quota aren't applied uniformly. They depend on when the user data is fetched from the LDAP. If you set a default quota value "A", then you change it to "B" and finally change it again to "C", quota values for different users might be "A", "B" or "C" at some point. It's expected that when you access to the user data (or the user logs in) the quota is applied correctly. However, keep in mind that there are some cases (check above) where the quota isn't changed, which will result in unexpected values for the quota.

As @jvillafanez said, it is needed to log in or show in the user's list the user in order to get the field configkey  set as 'quota' in the oc_preferences table.

**Comment by davitol** [comment]
BTW I think we should change the default values from B to MB. (I know that it is possible to set XXX MB and works) but IMO byte is too small to be considered as Quota default unit. I'll open an enhancement ticket for that.

File: lib/User/User.php [hunk 1]
```diff
@@ -460,23 +464,51 @@ public function updateQuota($valueFromLDAP = null) {
...
 
+		$quota = false;
 		if(is_null($valueFromLDAP)) {
 			$quotaAttribute = $this->connection->ldapQuotaAttribute;
 			if ($quotaAttribute !== '') {
 				$aQuota = $this->access->readAttribute($this->dn, $quotaAttribute);
 				if($aQuota && (count($aQuota) > 0)) {
-					$quota = $aQuota[0];
+					if ($this->verifyQuotaValue($aQuota[0])) {
+						$quota = $aQuota[0];
```
Comment (PVince81) [inline lib/User/User.php]: I'm confused here even when trying to read the above code.

Can you add PHPDoc to remove all doubts ?

Is `$valueFromLDAP` the value of quota attribute from the specific user read from LDAP ?
If yes, then what is the difference between `$valueFromLDAP` and the `ldapQuotaAttribute` value that is read at this line ?

File: lib/User/User.php [hunk 2]
```diff
@@ -460,23 +464,51 @@ public function updateQuota($valueFromLDAP = null) {
...
+					} else {
+						$this->log->log('not suitable LDAP quota found for user ' . $this->uid . ': [' . $aQuota[0] . ']', \OCP\Util::WARN);
+					}
 				}
 			}
+		} else {
+			if ($this->verifyQuotaValue($valueFromLDAP)) {
+				$quota = $valueFromLDAP;
```
Comment (PVince81) [inline lib/User/User.php]: which value is this ? the default or the one read from the user's quot attribute ? (add PHPDoc to remove all doubts)

File: lib/User/User.php [hunk 3]
```diff
@@ -460,23 +464,51 @@ public function updateQuota($valueFromLDAP = null) {
...
+
+		if ($quota === false) {
+			// quota not found using the LDAP attribute (or not parseable). Try the default quota
+			$defaultQuota = $this->connection->ldapQuotaDefault;
+			if ($this->verifyQuotaValue($defaultQuota)) {
+				$quota = $defaultQuota;
+			}
 		}
+
+		$targetUser = $this->userManager->get($this->uid);
```
Comment (PVince81) [inline lib/User/User.php]: I assume that calling this here is not causing any recursion ? See https://github.com/owncloud/user_ldap/issues/47 for the reason why I'm asking

File: tests/User/UserTest.php [hunk 4]
```diff
@@ -199,15 +199,17 @@ public function testUpdateQuotaAllProvided() {
...
+			->with($this->equalTo('ldapQuotaAttribute'))
+			->will($this->returnValue('myquota'));
 
+		/* Having a quota defined, the ldapQuotaDefault won't be used
 		$connection->expects($this->at(1))
 			->method('__get')
-			->with($this->equalTo('ldapQuotaAttribute'))
-			->will($this->returnValue('myquota'));
+			->with($this->equalTo('ldapQuotaDefault'))
+			->will($this->returnValue('23 GB'));
```
Comment (PVince81) [inline tests/User/UserTest.php]: did you comment out this test on purpose ?

File: tests/User/UserTest.php [hunk 5]
```diff
@@ -289,15 +291,17 @@ public function testUpdateQuotaIndividualProvided() {
...
+			->with($this->equalTo('ldapQuotaAttribute'))
+			->will($this->returnValue('myquota'));
 
+		/* Having a quota set this won't be used
 		$connection->expects($this->at(1))
 			->method('__get')
-			->with($this->equalTo('ldapQuotaAttribute'))
-			->will($this->returnValue('myquota'));
+			->with($this->equalTo('ldapQuotaDefault'))
+			->will($this->returnValue(''));
```
Comment (PVince81) [inline tests/User/UserTest.php]: same question here

File: lib/User/User.php (hunk 3 above)
Comment (PVince81) [inline lib/User/User.php]: I see we did it before, but now we do it in a new code path. I guess it shouldn't hurt.

**Comment by PVince81** [comment]
Missing test for the quota parser which now accepts "default" and "none" ?

**Comment by PVince81** [comment]
Looks good in general. I think this is great for 10.0, not sure about backporting as this would change behaviors or potentially cause surprises if some admins were using the current behavior in a legitimate way, if that's even possible. @jvillafanez thoughts on this ?

cc @butonic

File: lib/User/User.php (hunk 3 above)
Comment (jvillafanez) [inline lib/User/User.php]: There might be a one step recursion cut down by the `wasRefreshed` check.

I don't think there is a good solution for this. Either we hack the `setQuota` call and bypass the user management to set the quota directly in the DB (as it was done in previous versions) or we fetch the wrapped user from the core user manager to set the quota. I don't see any other solution without changing the whole architecture.

File: tests/User/UserTest.php (hunk 4 above)
Comment (jvillafanez) [inline tests/User/UserTest.php]: Yes, it was commented on purpose. Previously we were always fetching the default quota even though it wasn't being used in some scenarios. I think this way is a bit more clear that the default quota won't be fetched in this case
i'll check if I can do it properly.

Comment (jvillafanez) [inline tests/User/UserTest.php]: looks like I can't test that a method is never called with specific parameters... any other alternative?

File: lib/User/User.php (hunk 1 above)
Comment (jvillafanez) [inline lib/User/User.php]: PHPDoc updated with more information

File: lib/User/User.php (hunk 2 above)
Comment (jvillafanez) [inline lib/User/User.php]: PHPDoc updated

File: tests/User/UserTest.php (hunk 5 above)
Comment (jvillafanez) [inline tests/User/UserTest.php]: It's commented on purpose, to make clear that the LDAP default quota won't be fetched in that case.
It doesn't seem possible to check if the method won't be called with the specified parameters, so I'm not sure if leave this commented or remove it.

**Comment by jvillafanez** [comment]
I think the intention for the LDAP quota is more or less as explained in the PR, and everyone would expect the app to behave as explained (except for the corner cases, which I don't think we can solve). If the app isn't behaving as explained it should be considered as a bug.

There are already reports about strange quota behaviour, and this PR might solve them or at least help to debug what's happening with the quota.

**Comment by PVince81** [comment]
Looks good, thanks 👍 

For the commented out test I don't have any ideas right now.

**Comment by PVince81** [comment]
Please backport

**Comment by jvillafanez** [comment]
Backport to stable9.1 in https://github.com/owncloud/core/pull/27420
Any additional backport needed?

**Comment by PVince81** [comment]
If the behavior changed in OC 9 but was fine in OC 8.2 then we should likely also backport to stable9

**Comment by jvillafanez** [comment]
Backport for stable9 in https://github.com/owncloud/core/pull/27424

**Comment by IRAM38** [comment]
hello Everybody!
I have the same issue, I don't understand. I have Cloud Vers9.1.5 (stable)

Warning user_ldap not suitable default quota found for user xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

And the user Quota (for exemple 50Gb) is set to Default quota (4Gb) ..
Thanks for your help.

**Comment by michaelstingl** [comment]
@IRAM38 No, I don't think you have the same issue. This is a [pull request](https://help.github.com/articles/about-pull-requests/) and the related issue has been fixed in 9.1.5: https://github.com/owncloud/core/pull/27420

https://central.owncloud.org/ is a good starting point for help.
